Assignments/Sorting/QuickSort.playground/Pages/Quicksort In-place.xcplaygroundpage/Contents.swift

```diff
@@ -113,30 +143,45 @@ When the algorithm is running, there are 4 groups of elements. We need 3 indices
 */
 
 func dutchFlagPartition<T: Comparable>(inout array: [T], pivotIndex: Int, low: Int, high: Int) -> (Int, Int) {
-    let pivot = array[pivotIndex]
-    
-    var smaller = 0
-    var equal = 0
-    var larger = high
-    
-    // ...
-    
-    return (smaller, larger)
+  let pivot = array[pivotIndex]
+  
+  var smaller = 0
+  var equal = 0
+  var larger = high
+  
+  while equal <= larger {
+    if array[equal] < pivot {
+      if smaller != equal {
+        swap(&array[smaller], &array[equal])
+      }
+      smaller += 1
+      equal += 1
+    } else if array[equal] > pivot {
+      if equal != larger {
+        swap(&array[equal], &array[larger])
+      }
+      larger -= 1
+    } else {
+      equal += 1
```

Comment:
Did you notice the function running indefinitely here? I also implemented the wikipedia's pseudocode and smaller is always ++ while the function is not breaking/return the tuple.